R/utils.R

```diff
@@ -24,24 +24,64 @@ calculate_collimit = function(colwidths, outwidth) {
   collimit - 3  # subtracting 3 here because data.table adds "..." whenever it truncates a string
 }
 
-# same as task$filter(), but allows duplicate row IDs
+# Same as task$filter(), but allows duplicate row IDs.
+# Handles duplicate rows in tasks with col_role "group" by renaming groups
+# following pattern grp_name_1, grp_name_1, ... per each duplication of a group.
 # @param task [Task] the task
 # @param row_ids [numeric] the row IDs to select
 # @return [Task] the modified task
 task_filter_ex = function(task, row_ids) {
 
-  addedrows = row_ids[duplicated(row_ids)]
-
-  newrows = task$nrow + seq_along(addedrows)
+  # Get vector of duplicate row IDs
+  dup_ids = row_ids[duplicated(row_ids)]
+  # Generate vector with new row IDs
+  newrows = task$nrow + seq_along(dup_ids)
+  # Get all columns of task for subsetting using task$data()
+  cols = unique(unlist(task$col_roles, use.names = FALSE))
+
+  # Rbind duplicated rows to task
+  if (length(dup_ids)) {
+
+    # First, get a data.table with all duplicated rows.
+    new_data = task$data(rows = dup_ids, cols = cols)
+
+    # Second, if task has a column with role "group", create new groups for duplicate rows by adding a suffix to the group entry.
+    if (!is.null(task$groups)) {
+      group = NULL  # for binding
+      row_id = NULL  # for binding
+
+      # We create a data.table "new_groups" with the corresponding group to each duplicated ID.
+      #   1. Remove duplicates from task$groups which could exist in case of duplicates in task$row_roles$use.
+      #   2. Get correct number and positioning of rows through dup_ids.
+      # We then change the group entry based on how often the ID occurs. E.g. row_id = 1 occurs
+      # two times has the group entry "g". Then we rename the group entries to "g_1" and "g_2".
+      # If a group with a suffix (e.g. "_1") already exists, we add another suffix to it (i.e. "_1_1").
+      grps = unique(task$groups$group)
+      new_groups = unique(task$groups, by = "row_id")[J(dup_ids), on = "row_id"][, group := {
+        groups = character(0)
+        i = 1
+        while (length(groups) < .N) {
+          new_group = paste0(group[[1]], "_", i)
+          if (new_group %nin% grps) groups[[length(groups) + 1]] = new_group
+          i = i + 1
+        }
+        groups
+      }, by = row_id]
+
+      # Use "new_groups" to update the group entries.
+      new_data[, (task$col_roles$group) := new_groups$group]
```

Comment:
Test cases we want to check: Consider task with 3 rows, group 'a', 'b', 'c'; row roles "use" c(1, 1, 2, 2, 3)
* row_ids c(1, 1, 2, 2, 3) -> we get flattened task that has the same groups as before (has ids 1, 2, 3, 4, 5, groups 'a', 'b', 'c', 'a', 'b')
* row_ids c(1, 1, 1, 1, 2, 2, 3) -> add group a_1, i.e. task has c(1, 2, 3, 4, 5, 6, 7), groups 'a', 'b', 'c', 'a', 'b', 'a_1', 'a_1'
* row_ids c(1, 1, 1, 1, 3) -> same as above, but 'b' disappears: c(1, 2, 3, 4, 5), groups 'a', 'a', 'c', 'a_1', 'a_1'
* row_ids 3 -> groups a and b disappear: task has row-id 3, group 'c'

Reply:
* row_ids c(1, 1, 1, 1, 3, 1, 1) -> c(1, 3, 4, 5, 6, 7, 8), groups 'a', 'c', 'a', 'a_1', 'a_1', 'a_2', 'a_2'